lib/puppet/util/selinux.rb

```diff
@@ -47,8 +47,25 @@ def get_selinux_default_context(file)
     rescue Errno::EACCES, Errno::ENOENT
       mode = 0
     end
-
-    retval = Selinux.matchpathcon(file, mode)
+    # matchpathcon does only act on the passed string and not resolve the real path
+    # problematic if puppet manages /etc/init.d/script on RHEL which symlinks
+    # /etc/init.d to /etc/rc.d/init.d. matchpathcon will return the wrong label.
+    #
+    # This mimics restorecon behaviour.
+    #
+    # Not all files requested exist. Fall back to the requested file string.
+    begin
+      if !File.symlink?(file) && File.directory?(file)
+        file_realpath = File.realpath(file)
```

Comment:
How does the `File.realpath(file)` here differ from `File.realpath(File.dirname(file)) + '/' + File.basename(file)` performed in the else?  I would assume `File.realpath` resolves a file underneath a symlinked directory correctly, but perhaps that's not the case?